devices/webcam/hand_tracking_utility.py

```diff
@@ -0,0 +1,600 @@
+
+    @attributes:
+        log_images: bool, whether or not to log images to file
+        log_poses: bool, whether or not to log pose data to file
+        file_path: str, the path of the log file
+    """
+
+    log_images: bool
+    log_poses: bool
+    file_path: str
+
+class LoggingState(lg.State):
+    """
+    State object for data logging
+
+    @attributes:
+        file: h5py.File
+        closed: bool, true if the file is closed
+        frame_index: the current index of the frame received from the VideoFrame topic
+    """
+
+    file: Optional[h5py.File] = None
+    closed: bool = True
+    frame_index: int = 0
+
+class Logging(lg.Node):
+    """
+    Data logging node
+
+    @topics:
+        INPUT: PoseFrame
+    
+    @attributes:
+        state: LoggingState
+        config: LoggingConfig
+    """
+
+    INPUT = lg.Topic(PoseFrame)
+    state: LoggingState
+    config: LoggingConfig
+
+    @lg.subscriber(INPUT)
+    def process(self, message: PoseFrame) -> None:
+        if not self.state.closed and (self.config.log_images or self.config.log_poses):
+            if self.config.log_images:
+                self.state.file.create_dataset("images/{}/image".format(self.state.frame_index), data = message.frame)
+
+            if self.config.log_poses:
+                hand_index = 0
+                for landmark_list in message.landmarks:
+                    arr = np.zeros(shape = (len(landmark_list.landmark), 3), dtype = np.float32)
+                    for id, landmark in enumerate(landmark_list.landmark):
+                        arr[id] = [landmark.x, landmark.y, landmark.z]
+                    self.state.file.create_dataset("images/{0}/hand_landmarks/{1}".format(self.state.frame_index, hand_index), data = arr)
+                    hand_index += 1
+
+            self.state.frame_index += 1
+
+    def setup(self) -> None:
+        if (self.config.log_images or self.config.log_poses):
+            self.state.file = h5py.File(self.config.file_path, "a")
+            self.state.closed = False
+    
+    def cleanup(self) -> None:
+        if (self.config.log_images or self.config.log_poses):
+            self.state.closed = True
+            self.state.file.close()
+
+class LogStreamConfig(lg.Config):
+    """
+    Config object for streaming logged images
+
+    @attributes:
+        sample_rate: int, how many times per second to dispatch an image
+        file_path: str, path to the log file
+    """
+
+    sample_rate: int
+    file_path: str
+
+class LogStream(lg.Node):
+    """
+    Streams the VideoFrame topic from a log file
+
+    @topics:
+        OUTPUT: VideoFrame
+    
+    @attributes:
+        config: LogStreamConfig
+        state: LoggingState
+    """
+
+    OUTPUT = lg.Topic(VideoFrame)
+    config: LogStreamConfig
+    state: LoggingState
+
+    @lg.publisher(OUTPUT)
+    async def read_file(self) -> lg.AsyncPublisher:
+        while not self.state.closed:
+            start_time_ns = time.time_ns()
```

Comment:
For windows OS, perf_counter has a higher resolution than time_ns.
Reference: https://peps.python.org/pep-0564/#annex-clocks-resolution-in-python

Reply:
The latest commits now make use of `perf_counter` over `time_ns`.